src/librustc/middle/effect.rs

```diff
@@ -218,6 +222,25 @@ impl<'a, 'tcx> Visitor<'tcx> for EffectCheckVisitor<'a, 'tcx> {
                     }
                 }
             }
+            hir::ExprAssign(ref lhs, ref rhs) => {
+                if let hir::ExprField(ref base_expr, field) = lhs.node {
+                    if let ty::TyAdt(adt, ..) = self.tables.expr_ty_adjusted(base_expr).sty {
+                        if adt.is_union() {
+                            let field_ty = self.tables.expr_ty_adjusted(lhs);
+                            let owner_def_id = self.tcx.hir.body_owner_def_id(self.body_id);
+                            let param_env = self.tcx.parameter_environment(owner_def_id);
+                            if field_ty.moves_by_default(self.tcx, &param_env, field.span) {
```

Comment:
(I think what I would rather do is to have a loop over all the bodies and instantiate a distinct visitor per body, with just the appropriate tables etc. This visitor would have to walk up to find out contextual information though (since closures, at least, inherit their "unsafety" from their parents at present.))